Generate pseudo-class/element enum classes and enablement checks

nt1m opened the pull request:
#### 5ddbccf5974982d2fc1902c78dadffb0c98b4afd
<pre>
Generate pseudo-class/element enum classes and enablement checks
<a href="https://bugs.webkit.org/show_bug.cgi?id=266952">https://bugs.webkit.org/show_bug.cgi?id=266952</a>
<a href="https://rdar.apple.com/120354695">rdar://120354695</a>

Reviewed by Antti Koivisto.

This commit:
- Generates the PseudoClass/PseudoElement enum classes in CSSSelectorEnums.h
- Adds settings-flag data into CSSPseudoSelectors.json and uses it to generate methods in CSSSelectorInlines.h
- Starts making use of those files

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElement):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumePseudo):
* Source/WebCore/css/process-css-pseudo-selectors.py:
(write_copyright_header):
(key_is_true): Deleted.
(expand_ifdef_condition): Deleted.
(format_name_for_enum_class): Deleted.
(format_name_for_enum_class.format): Deleted.
(GPerfMappingGenerator): Deleted.
(GPerfMappingGenerator.__init__): Deleted.
(GPerfMappingGenerator.is_pseudo_selector_enabled): Deleted.
(GPerfMappingGenerator.generate_pseudo_class_and_compatibility_element_map): Deleted.
(GPerfMappingGenerator.generate_pseudo_class_and_compatibility_element_map.in): Deleted.
(GPerfMappingGenerator.generate_pseudo_element_map): Deleted.
(GPerfOutputGenerator): Deleted.
(GPerfOutputGenerator.__init__): Deleted.
(GPerfOutputGenerator.write_copyright_header): Deleted.

Canonical link: <a href="https://commits.webkit.org/272565@main">https://commits.webkit.org/272565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/33adf359f89aae4d9f61c0021cbcd004005fb519

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32138 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33941 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34692 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29131 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32979 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8080 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28710 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32549 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9164 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28728 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7966 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8129 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28643 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36037 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29224 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29100 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34249 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8250 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6195 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32107 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9889 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7504 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8889 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8780 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -2544,16 +2544,19 @@ list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/CSSValueKeywords.cpp)
     VERBATIM)
 list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/SelectorPseudoClassAndCompatibilityElementMap.cpp)
 list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/SelectorPseudoElementMap.cpp)
+ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/CSSSelector.cpp CSSSelectorEnums.h)
+ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/CSSSelector.cpp CSSSelectorInlines.h)
+ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/CSSSelectorParser.cpp CSSSelectorInlines.h)
```
@mcatanzaro I'm not sure if this is needed for the GTK port (nor if it's the right thing), but I added it anyway.

mcatanzaro replied:
This means "rebuild CSSSelectorEnums.h and CSSSelectorInlines.h when CSSSelector.cpp changes, and rebuild CSSSelectorInlines.h when CSSSelectorParser.cpp changes." And it looks like that is what you want, but a better way to write it would be to add a `DEPENDS` to the add_custom_command() above. I will submit a follow-up for this.

I actually don't see any reason for this macro to exist at all, so I'll probably remove it.

mcatanzaro replied:
https://bugs.webkit.org/show_bug.cgi?id=266984

annevk commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -376,63 +389,265 @@ def write_parsing_function_definitions_for_pseudo_element(self, writer):
+
+            # End switch statement.
+            writer.write("}")
+            writer.write("return true;")
```
Nit: ASSERT_NOT_REACHED?

nt1m replied:
C++ doesn't require you to return if the function only has a switch statement, so I guess I'll just omit the return, like in: https://searchfox.org/wubkat/source/__GENERATED__/WebCore/DerivedSources/CSSPropertyParsing.cpp#195-206

mcatanzaro replied:
It's fine because you have a `default` case, so the function is guaranteed to return and there's no point in adding an assert.

(If it were missing the `default` case, then the Linux EWS would have failed on -Werror=return-type.)

annevk commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -376,63 +389,265 @@ def write_parsing_function_definitions_for_pseudo_element(self, writer):
+
+            # End switch statement.
+            writer.write("}")
+            writer.write("return true;")
```
Nit: ASSERT_NOT_REACHED?

nt1m replied:
ditto

annevk reviewed (commented):
The one thing that's still a bit complex with settings here is that there's a bunch of manual work still to get the setting exposed here, but hopefully we can eventually improve that as well.

Looks very nice overall.

nt1m committed:
Generate pseudo-class/element enum classes and enablement checks
https://bugs.webkit.org/show_bug.cgi?id=266952
rdar://120354695

Reviewed by Antti Koivisto.

This commit:
- Generates the PseudoClass/PseudoElement enum classes in CSSSelectorEnums.h
- Adds settings-flag data into CSSPseudoSelectors.json and uses it to generate methods in CSSSelectorInlines.h
- Starts making use of those files

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElement):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumePseudo):
* Source/WebCore/css/process-css-pseudo-selectors.py:
(write_copyright_header):
(key_is_true): Deleted.
(expand_ifdef_condition): Deleted.
(format_name_for_enum_class): Deleted.
(format_name_for_enum_class.format): Deleted.
(GPerfMappingGenerator): Deleted.
(GPerfMappingGenerator.__init__): Deleted.
(GPerfMappingGenerator.is_pseudo_selector_enabled): Deleted.
(GPerfMappingGenerator.generate_pseudo_class_and_compatibility_element_map): Deleted.
(GPerfMappingGenerator.generate_pseudo_class_and_compatibility_element_map.in): Deleted.
(GPerfMappingGenerator.generate_pseudo_element_map): Deleted.
(GPerfOutputGenerator): Deleted.
(GPerfOutputGenerator.__init__): Deleted.
(GPerfOutputGenerator.write_copyright_header): Deleted.

Canonical link: https://commits.webkit.org/272565@main
